### PR TITLE
Update of mkFit for 12_1_0_pre4

### PR DESCRIPTION
**Pull request by mmasciov**
#### PR description:

This PR follows #35199 and updates mkFit in view of CMSSW_12_1_0_pre4.

In detail, this PR includes technical update of mkFit configurations.
Via  cms-sw/cmsdist#7349 and cms-data/RecoTracker-MkFit#5 it includes updates to improve tracking performance, presented in detail at the [Tracker DPG - Tracking POG general meeting](https://indico.cern.ch/event/992975/?showDate=all&showSession=all#8-mkfit-status-and-plans).

It requires cms-sw/cmsdist#7349 and cms-data/RecoTracker-MkFit#5

#### PR validation:

Performance has been presented in detail at the [Tracker DPG - Tracking POG general meeting](https://indico.cern.ch/event/992975/?showDate=all&showSession=all#8-mkfit-status-and-plans)

Full MTV results in TTbar events with PU (Run-3, 2021) are available:
http://uaf-10.t2.ucsd.edu/~mmasciov/MkFit_devs/TrackerDPGValidation/MTV_mkFit-7iter_pre4_plusCKF_Sep27/

### CODE REVIEW

**Comment by makortel** [comment]
code-checks with cms.week0.PR_bd16d39b/52.0-e05c6c4771494e730b9d4ac54b12b03d

**Comment by makortel** [comment]
test parameters:
   - pull_requests = cms-sw/cmsdist#7349,cms-data/RecoTracker-MkFit#5

**Comment by makortel** [comment]
@cmsbuild, please test

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc [hunk 1]
```diff
@@ -6,35 +6,107 @@
...
 #include "mkFit/IterationConfig.h"
 
+namespace {
+  using namespace mkfit;
+
+  void partitionSeeds0(const TrackerInfo &trk_info,
+                       const TrackVec &in_seeds,
+                       const EventOfHits &eoh,
+                       IterationSeedPartition &part) {
+    const int size = in_seeds.size();
```
Comment (jpata) [inline RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc:21]:
```suggestion
    const size_t size = in_seeds.size();
```
perhaps?

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc [hunk 2]
```diff
@@ -6,35 +6,107 @@
...
+
+    for (int i = 0; i < size; ++i) {
```
Comment (jpata) [inline RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc:23]:
```suggestion
    for (size_t i = 0; i < size; ++i) {
```

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc [hunk 3]
```diff
@@ -6,35 +6,107 @@
...
+      const Track &S = in_seeds[i];
+
+      const bool z_dir_pos = S.pz() > 0;
+
+      HitOnTrack hot = S.getLastHitOnTrack();
+      float eta = eoh[hot.layer].GetHit(hot.index).eta();
```
Comment (jpata) [inline RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc:29]:
```suggestion
      const float eta = eoh[hot.layer].GetHit(hot.index).eta();
```

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc [hunk 4]
```diff
@@ -6,35 +6,107 @@
...
+
+      const LayerInfo &tib1 = trk_info.m_layers[4];
+      const LayerInfo &tob1 = trk_info.m_layers[10];
+
+      const LayerInfo &tecp1 = trk_info.m_layers[27];
+      const LayerInfo &tecn1 = trk_info.m_layers[54];
+
+      const LayerInfo &tec_first = z_dir_pos ? tecp1 : tecn1;
+
+      float maxR = S.maxReachRadius();
```
Comment (jpata) [inline RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc:44]:
```suggestion
      const float maxR = S.maxReachRadius();
```

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc [hunk 5]
```diff
@@ -6,35 +6,107 @@
...
+        if ((S.canReachRadius(tib1.m_rin) && tib1.is_within_z_limits(S.zAtR(tib1.m_rin))) ||
+            (S.canReachRadius(tob1.m_rin) && tob1.is_within_z_limits(S.zAtR(tob1.m_rin)))) {
+          reg = z_dir_pos ? TrackerInfo::Reg_Transition_Pos : TrackerInfo::Reg_Transition_Neg;
+        } else {
+          reg = z_dir_pos ? TrackerInfo::Reg_Endcap_Pos : TrackerInfo::Reg_Endcap_Neg;
+        }
+      }
+
+      part.m_region[i] = reg;
+      part.m_sort_score[i] = 5.0f * (reg - 2) + eta;
```
Comment (jpata) [inline RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc:76]: perhaps these magic numbers need to be put into a constexpr or clarified via a comment?

**Review by jpata — COMMENTED** [review]
Are the differences in the mkFit / .7 workflows as expected?

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc (hunk 1 above)
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc]: Done, thanks.

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc (hunk 2 above)
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc]: Done, thanks.

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc (hunk 3 above)
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc]: Done, thanks.

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc (hunk 4 above)
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc]: Done, thanks.

**Comment by mmasciov** [comment]
> Are the differences in the mkFit / .7 workflows as expected?

@jpata Yes, minor changes in tracking performance are expected, as presented / discussed at the last [Tracker DPG - Tracking POG general meeting](https://indico.cern.ch/event/992975/?showDate=all&showSession=all#8-mkfit-status-and-plans).

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
Just to note, I'm ready to sign this PR, I'm waiting for https://github.com/cms-data/RecoTracker-MkFit/pull/5 and https://github.com/cms-sw/cmsdist/pull/7349 to be merged before leaving the reco sig.

**Comment by slava77** [comment]
> Just to note, I'm ready to sign this PR, I'm waiting for [cms-data/RecoTracker-MkFit#5](https://github.com/cms-data/RecoTracker-MkFit/pull/5) and [cms-sw/cmsdist#7349](https://github.com/cms-sw/cmsdist/pull/7349) to be merged before leaving the reco sig.

in this case all 3 PRs have to be included at the same time. So, serializing the signature will not really work.

**Comment by jpata** [comment]
+reconstruction
- updates the mkFit version
- changes expected/observed in the mkFit/.7 workflows
- requires cms-data/RecoTracker-MkFit#5 and cms-sw/cmsdist#7349

**Comment by perrotta** [comment]
+1
- Merged for 12_1_0_pre4, as agreed right now at the ORP meeting

**Pull request merged**